### PR TITLE
Fix CMSSW seg fault showing up as BadFWJRXML error

### LINKED ISSUE #9289: CMSSW seg fault showing up as BadFWJRXML error in WMStats
Provided that jobs can pile many different exceptions and exit codes within the same step, I'd expect the very first exception to be reported in WMStats.

Which is not the case for (in testbed)
amaltaro_TaskChain_NanoAOD_Agent124_Val_190706_191332_1418

there is a job that seg faulted with an exit code 139, followed then by a failure to read the XML job report with an exit code 50115 (of course, the report wasn't completely created). What is getting reported in the WMStats table is actually the 50115 first, followed by the other 2 exit codes added to the job.
```
Job Name: 304435bf-e897-410b-bc6e-a8a7e6771653-0
WMBS job id: 39
Workflow: amaltaro_TaskChain_NanoAOD_Agent124_Val_190706_191332_1418
Task: /amaltaro_TaskChain_NanoAOD_Agent124_Val_190706_191332_1418/RunIIFall17NanoAOD
Status: jobfailed
Input dataset: /WJetsToLNu_HT-200To400_TuneCUETP8M1_13TeV-madgraphMLM-pythia8/RunIISummer16MiniAODv2-PUMoriond17_80X_mcRun2_asymptotic_2016_TrancheIV_v6_ext1-v1/MINIAODSIM
Site: T2_CH_CERN
Agent: vocms0193.cern.ch
ACDC URL: https://cmsweb-testbed.cern.ch/couchdb/acdcserver
State Transition
Exit code: 50115
Retry count: 0
Errors:

    logArch1

    cmsRun1
        BadFWJRXML (Exit Code: 50115)

            Error reading XML job report file, possibly corrupt XML File:
            Details: no element found: line 1, column 0

        CMSSWStepFailure (Exit Code: 139)

             Adding last 25 lines of CMSSW stdout:
            #3  <signal handler called>
```

Given that 139 was the first error, it should be the first error to be added to that step. As well as the first error to get reported in the WMStats table.

So we need to investigate:
* whether the order of errors are correct in the job report
* and why wmstats is changing that order (or which order/logic it's using).

Job log can be found here: http://amaltaro.web.cern.ch/amaltaro/forWMCore/Issue_9289/

### PR DESCRIPTION
**Pull request by khurtado**
Fixes #9289

#### Status
Tested and working. 

#### Description
Whenever the method parse() in FwkJobReport is used, it adds an error to the report when it fails.
We use this method in Executors/CMSSW and Diagnostics/CMSSW. The latter adds the errors corresponding to cmsRun exit code, and also bad job report errors,etc. However, since we also call this funcion from executors and this runs first, we end up adding the BadFWJRXML error report first, before anything coming from Diagnostics/CMSSW.

This change adds an option to parse(), so that addError can be disabled when needed. This is enabled by default, but it is disabled when called from Executors/CMSSW.

#### Is it backward compatible (if not, which system it affects?)
YES

### CODE REVIEW

**Comment by khurtado** [comment]
test this please

**Comment by khurtado** [comment]
test this please

**Comment by khurtado** [comment]
test this please

**Comment by amaltaro** [comment]
@khurtado don't bother about the Rucio unit tests. They were supposed to be marked as unstable tests at the moment, I haven't figured out yet why Jenkins don't do so.

**Review by amaltaro — CHANGES_REQUESTED** [review]
Kenyi, as we discussed on the WMCore meeting, I think we completely remove the `addError` from the `parse` method and simply rely on the Diagnostics package. Where this `50115` exit code is disabled for the CMSSW step.

Honestly speaking, I'm not sure how useful it's to add/propagate this bad FJR, given that we already add the original job exception and error message..

Can you please make that change, run a workflow that always crash a job or two (and also condor_rm another job or two) and see how the errors get reported?

**Comment by khurtado** [comment]
@amaltaro To be sure I understand, do you mean explicitly calling the `executor.report.addError` in all Diagnostic classes that call `parse()`, like [CMSSW.py](https://github.com/dmwm/WMCore/blob/a9026342eba03286bc510512211ca2913b57d80a/src/python/WMCore/WMSpec/Steps/Diagnostics/CMSSW.py), [DeleteFiles.py](https://github.com/dmwm/WMCore/blob/0615bd977edd4ad103db7ea24f461aa794c9a104/src/python/WMCore/WMSpec/Steps/Diagnostics/DeleteFiles.py#L43), [LogArchive.py](https://github.com/dmwm/WMCore/blob/0615bd977edd4ad103db7ea24f461aa794c9a104/src/python/WMCore/WMSpec/Steps/Diagnostics/LogArchive.py#L45), [AlcaHarverst.py](https://github.com/dmwm/WMCore/blob/0615bd977edd4ad103db7ea24f461aa794c9a104/src/python/WMCore/WMSpec/Steps/Diagnostics/AlcaHarvest.py#L48), etc?

**Comment by amaltaro** [comment]
I don't remember exactly where I got when I was looking at the code, but yes, I think that's what I meant. However, we don't need to add this `addError` call to every single error class, I think only to the default one (or something like that). 

Anyways, if you make such change, it's probably a better idea to keep it in a different commit for now.

**Comment by khurtado** [comment]
@amaltaro I removed the `addError` call from `Report` and made a new `parse` method in `Diagnostics` that calls the method from `Report` and adds the  `addError`. I didn't want to completely remove the `BadFWJRXML` error type, because even though it is obvious an xml report won't read if cmsRun segfaults and we get a Linux error code, I guess you can still have cmsRun finish with some failure and have the XML corrupted for some reason (IO error, full disk to write the file?)

Anyway, here is a task test:
Task name:     `wmagent_TaskChain_NanoAOD_khurtado_nanoaod_v21_190823_030849_8268`

1) exitCode shows 139 now (coming from CMSSWStepFailure)
2) Error order looks fine in the `Errors` section: `CMSSWStepFailure(139), BadFWJRXML(50115), ErrorLoggingAddition(99999)
3) Somehow, the error message tab shows `BadFWJRXML,CMSSWStepFailure,ErrorLoggingAddition`
I have no idea what the reason for this is, but I think 1 and 2 are the important ones?
4) Jobs deleted after they were running for a while report `NoJobReport (Exit Code: 99303)`
5) I still need to do more pylint fixes, let me know if you are ok with this approach and I can do those fixes.

```
Errors:

    logArch1

    cmsRun1
        CMSSWStepFailure (Exit Code: 139)

             Adding last 25 lines of CMSSW stdout:
 <snip>
            A fatal system signal has occurred: segmentation violation
            Complete
            process id is 228 status is 139
            <@========== WMException Start ==========@>
            Exception Class: CmsRunFailure
            Message: Error running cmsRun
            {'arguments': ['/bin/bash', '/srv/job/WMTaskSpace/cmsRun1/cmsRun1-main.sh', '', 'slc6_amd64_gcc630', 'scramv1', 'CMSSW', 'CMSSW_9_4_2', 'FrameworkJobReport.xml', 'cmsRun', 'PSet.py', '', '', '']}
            Linux Return code: 139

            	ModuleName : WMCore.WMSpec.Steps.WMExecutionFailure
            	MethodName : __init__
            	ClassInstance : None
            	FileName : /srv/job/WMCore.zip/WMCore/WMSpec/Steps/WMExecutionFailure.py
            	ClassName : None
            	LineNumber : 18
            	ErrorNr : 139

            Traceback: 
              File "/srv/job/WMCore.zip/WMCore/WMSpec/Steps/Executors/CMSSW.py", line 270, in execute
                self.report.parse(jobReportXML, stepName=self.stepName)

              File "/srv/job/WMCore.zip/WMCore/FwkJobReport/Report.py", line 156, in parse
                raise FwkJobReportException(msg)

            <@---------- WMException End ----------@>

        BadFWJRXML (Exit Code: 50115)

            Error reading XML job report file, possibly corrupt XML File:
            Details: <@========== WMException Start ==========@>
            Exception Class: FwkJobReportException
            Message: Error reading XML job report file, possibly corrupt XML File:
            Details: no element found: line 1, column 0
            	ModuleName : WMCore.FwkJobReport.Report
            	MethodName : parse
            	ClassInstance : None
            	FileName : /srv/job/WMCore.zip/WMCore/FwkJobReport/Report.py
            	ClassName : None
            	LineNumber : 156
            	ErrorNr : 0

            Traceback: 
              File "/srv/job/WMCore.zip/WMCore/FwkJobReport/Report.py", line 144, in parse
                xmlToJobReport(self, xmlfile)

              File "/srv/job/WMCore.zip/WMCore/FwkJobReport/XMLParser.py", line 503, in xmlToJobReport
                node = xmlFileToNode(xmlFile)

              File "/srv/job/WMCore.zip/WMCore/Algorithms/ParseXMLFile.py", line 67, in xmlFileToNode
                build(node))

              File "/srv/job/WMCore.zip/WMCore/Algorithms/ParseXMLFile.py", line 88, in expat_parse
                parser.ParseFile(f)

            <@---------- WMException End ----------@>

        ErrorLoggingAddition (Exit Code: 99999)

            Adding extra error in order to hold error report

            Adding last ten lines of CMSSW stderr:
            WARNING: In non-interactive mode release checks e.g. deprecated releases, production architectures are disabled.
            WARNING: There already exists /srv/job/WMTaskSpace/cmsRun1/CMSSW_9_4_2 area for SCRAM_ARCH slc6_amd64_gcc630.
            /srv/job/WMTaskSpace/cmsRun1/cmsRun1-main.sh: line 81:   228 Segmentation fault      (core dumped) $EXECUTABLE -j $JOB_REPORT $CONFIGURATION 2>&1
```

File: src/python/WMCore/WMSpec/Steps/Executor.py [hunk 1]
```diff
@@ -74,6 +74,7 @@ def __init__(self):
         self.workload = None
         self.job = None
         self.errorDestination = None
+        self.failedPreviousStep = None
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/Steps/Executor.py]: Isn't this only needed for the CMSSW module? If so, then I'd suggest to remove it from here.

**Review by amaltaro — COMMENTED** [review]
Kenyi, changes look Ok. 
However, I'm not really sure whether we can simply replace the `Report.parse` call by `Diagnostic.parse`, reason being that the Report.parse calls this method `xmlToJobReport`, which should be converting the CMSSW FWJR to a WMCore/Job FWJR.

Have you actually ran a workflow until the very end and checked whether data was pushed to DBS/PhEDEx? It would be one way to see whether this change in methods have a bad impact or not.

File: src/python/WMCore/WMSpec/Steps/Executor.py (hunk 1 above)
Comment (khurtado) [reply src/python/WMCore/WMSpec/Steps/Executor.py]: Yes, but pylint complained this wasn't declared in `__init__` as a "must fix" item.

Comment (amaltaro) [reply src/python/WMCore/WMSpec/Steps/Executor.py]: Don't trust everything pylint says ;) AFAIR, one classical case where it gets lost is when we set object attributes with `setattr`

Comment (khurtado) [reply src/python/WMCore/WMSpec/Steps/Executor.py]: regarding waiting to have a completed job, I haven't yet, still waiting :). I just injected a couple of extra different workflows that I know usually finish with no errors to make sure things are good.
Regarding the pylint thing... any idea on how to get rid of  `self.failedPreviousStep` in `__init__` and not have the pylint report as "failed"? Unless the failed report doesn't matter for this particular case.

Comment (amaltaro) [reply src/python/WMCore/WMSpec/Steps/Executor.py]: We could disable it, but I usually prefer to have these complaints than disable things.
BTW, is it complaining in the jenkins test or locally? I just ran pylint (1.9.4) on my laptop and didn't see such complain.

Comment (khurtado) [reply src/python/WMCore/WMSpec/Steps/Executor.py]: @amaltaro I can't make DAS queries via cmsweb-testbed.cern.ch, but PhEDEX is showing 
 
[test1](https://cmsweb-testbed.cern.ch/phedex/prod/Data::Replicas?bexp=11921801;filter=%20%2FRelValSinglePiE50HCAL%2FDMWM-TC_SLC7_khurtado_slc7test_cv1-v11%2FGEN-SIM;dexp=1420163#b11921801
)

[GEN-SIM in test2](/RelValProdMinBias/DMWM-StepChain_DupOutMod_khurtado_stepchaindup_v9-v11/GEN-SIM)

Comment (amaltaro) [reply src/python/WMCore/WMSpec/Steps/Executor.py]: Good! I guess I haven't mentioned, but we have this script doing some reasonable checks for us:
https://github.com/dmwm/WMCore/wiki/Injecting,-assigning-and-validating-test-requests#validating-requests

File: src/python/WMCore/WMSpec/Steps/Diagnostics/AlcaHarvest.py [hunk 2]
```diff
@@ -66,4 +66,5 @@ def __init__(self):
         self.handlers[60319] = Exit60319()
 
         catchAll = AHExceptionHandler()
-        [self.handlers.__setitem__(x, catchAll) for x in range(0, 255) if x not in self.handlers]
+        _ = [self.handlers.__setitem__(x, catchAll) for x in range(0, 255) if x
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/Steps/Diagnostics/AlcaHarvest.py]: Kenyi, please do the same here. Expand this list comprehension in 3 lines.

File: src/python/WMCore/WMSpec/Steps/Diagnostics/CMSSW.py [hunk 3]
```diff
@@ -294,4 +295,5 @@ def __init__(self):
         # for all the exception codes between 1 and 225, use a default that attempts to read the code
         # from the job report
         catchAll = EDMExceptionHandler()
-        [self.handlers.__setitem__(x, catchAll) for x in range(0, 255) if x not in self.handlers]
+        _ = [self.handlers.__setitem__(x, catchAll) for x in range(0, 255) if x
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/Steps/Diagnostics/CMSSW.py]: same here

File: src/python/WMCore/WMSpec/Steps/Diagnostics/DeleteFiles.py [hunk 4]
```diff
@@ -64,4 +63,5 @@ def __init__(self):
 
 
         catchAll = DFExceptionHandler()
-        [ self.handlers.__setitem__(x, catchAll) for x in range(0, 255) if x not in self.handlers ]
+        _ = [self.handlers.__setitem__(x, catchAll) for x in range(0, 255) if x
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/Steps/Diagnostics/DeleteFiles.py]: same here

File: src/python/WMCore/WMSpec/Steps/Diagnostics/LogArchive.py [hunk 5]
```diff
@@ -72,4 +71,5 @@ def __init__(self):
         catchAll            = LAExceptionHandler()
         self.defaultHandler = catchAll
 
-        [ self.handlers.__setitem__(x, catchAll) for x in range(0, 255) if x not in self.handlers ]
+        _ = [self.handlers.__setitem__(x, catchAll) for x in range(0, 255) if x
```
Comment (amaltaro) [inline src/python/WMCore/WMSpec/Steps/Diagnostics/LogArchive.py]: same here

**Review by amaltaro — CHANGES_REQUESTED** [review]
There are a few more places. Please also squash real fix (src/ changes) from the test/pylint into 2 different commits.

File: src/python/WMCore/WMSpec/Steps/Diagnostics/AlcaHarvest.py (hunk 2 above)
Comment (amaltaro) [reply src/python/WMCore/WMSpec/Steps/Diagnostics/AlcaHarvest.py]: Sorry Kenyi, I meant this:
```
for x in range(0, 255):
    if x not in self.handlers
        self.handlers.__setitem__(x, catchAll)
```

**Comment by amaltaro** [comment]
Patch looks good, Kenyi.
Indeed I see a different ordering in the wmstats table, even though the job report file contains the correct order of errors, e.g.:
```
>>> for errorNum, msg in jobReport.getStepErrors("cmsRun2").iteritems():
    print("**** Error message for: %s" % errorNum)
    print(msg.dictionary_whole_tree_())
**** Error message for: error0
{'type': 'CMSSWStepFailure', 'details': '\n Adding last 25 lines of CMSSW stdout:\n%MSG-w MemoryCheck:  HGCalBackendLayer1Producer:hgcalBackEndLayer...
**** Error message for: error1
{'type': 'BadFWJRXML', 'details': 'Error reading XML job re...
```

We might be doing something on the visualization layer which orders the errors, or just not fetching them in order. I'll see if I can find what's wrong, otherwise we just proceed with this patch.

**Review by amaltaro — APPROVED** [review]

**Pull request merged**